### DISCUSSION TITLE
Why `Reconcile` function is called many times in `controller_test.go` ?

### DISCUSSION THREAD

**Discussion by LittleWat** [kubernetes-sigs/kubebuilder#2836] [Support] [answered]
Hello. I am new to kubebuilder.

Now I am implementing a controller_test that verifies the mock function called `Once` using testify.

But the `Reconcile` function is called many times even though I return `&ctrl.Result{Requeue: false}` in  `Reconcile` function.

Is it possible to call `Reconcile` function once in test codes...?

**Comment by camilamacedo86** [comment, marked as answer]
Hi @LittleWat, 

By using the Operators, it’s possible not only to provide all expected resources but also to manage them dynamically, programmatically, and at runtime. To illustrate this idea, imagine if someone accidentally changed a configuration or removed a resource by mistake; in this case, the Operator could fix it without any human intervention.  

**Therefore, the goal is to always ensure a desired state on the cluster via idempotent solutions, and because of this, we "watch" the required resources on the cluster and when something changes we reconcile.** 

> TL'DR: I recommend you check the common suggestions doc under the best practices section on the SDK website, see: https://sdk.operatorframework.io/docs/best-practices/common-recommendation/. for further info

Regards your question:

> Now I am implementing a controller_test that verifies the mock function called Once using testify.

Not sure how you are doing this implementation but a good example can be found in the scaffold done by the new plugin [deploy.image/v1-alpha](https://master.book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html), see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3-with-deploy-image/controllers/busybox_controller_test.go#L36-L125

> But the Reconcile function is called many times even though I return &ctrl.Result{Requeue: false} in Reconcile function.

Again not sure how you are doing that. If it is called more than once or it is returning errors which will make the reconciliation begin again or the resources which are watched have been changed and then are calling the reconciliation again. 

Also, see the blog: https://dev4devs.com/2020/08/16/how-to-getting-started-develop-go-operators-from-scratch-with-sdk-1-0/ it has good info about getting started and describes the returns for the reconcile.

**Reply by LittleWat** [reply]
Thank you for your quick and detailed answer!!!
This helps me a lot.

>https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3-with-deploy-image/controllers/busybox_controller_test.go#L36-L125

Ah, I misunderstood that I had to create `k8sManager` like below and  that `Reconcile` function should be called from the manager. 
```
	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
		Scheme: scheme.Scheme,
	})
	Expect(err).ToNot(HaveOccurred())
```
But it was not necessary. Just calling `Reconcile`  function satisfies what I wanted to do!

**Reply by camilamacedo86** [reply]
That is necessary to pass the API Schemas to the mock manager.

**Reply by LittleWat** [reply]
Yes, I misunderstood that kubebuilder tests always need the step to create a manager and set up a reconsiler with the manager like the sample code below:

https://book.kubebuilder.io/cronjob-tutorial/writing-tests.html
```
    k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
        Scheme: scheme.Scheme,
    })
    Expect(err).ToNot(HaveOccurred())

    err = (&CronJobReconciler{
        Client: k8sManager.GetClient(),
        Scheme: k8sManager.GetScheme(),
    }).SetupWithManager(k8sManager)
    Expect(err).ToNot(HaveOccurred())
``` 

Anyway, I could have solved this issue, and really appreciate your help! 🙇

**Comment by camilamacedo86** [comment]
Closing since it is answered.

**Reply by LittleWat** [reply]
thank you!
